File: Strivers-A2Z-DSA-Sheet/05.Recursion/2.Subsequences Pattern/javascript/01.Generate all valid parenthesis.js
/*
=============================================================================
  QUESTION: 22. Generate Parentheses (LeetCode)
=============================================================================

  Given n pairs of parentheses, generate all valid (well-formed) combinations.

  Examples:
    n = 1 → ["()"]
    n = 3 → ["((()))","(()())","(())()","()(())","()()()"]

=============================================================================
  APPROACH: Recursion with Open/Close Count — O(2^N × N) Time, O(N) Space
=============================================================================

  Rules:
  - Add '(' if open < n
  - Add ')' only if close < open (can't close what isn't opened)
  - When both open and close == n → valid string, push to result

  Dry Run (n=2):
    ""  →  "("  →  "(("  →  "(()"  →  "(())" ✓
                →  "()"  →  "()("  →  "()()" ✓

=============================================================================
*/

function generateParenthesis(n) {
    let result = [];

    function solve(s, open, close) {
        if (open === n && close === n) {
            result.push(s);
            return;
        }
        if (open < n) solve(s + '(', open + 1, close);
        if (close < open) solve(s + ')', open, close + 1);
    }

    solve("", 0, 0);
    return result;
}

// ==========================================================================
// DRIVER CODE
// ==========================================================================
console.log(generateParenthesis(1));  // ["()"]
console.log(generateParenthesis(2));  // ["(())","()()"]
console.log(generateParenthesis(3));  // ["((()))","(()())","(())()","()(())","()()()"]

File: Strivers-A2Z-DSA-Sheet/05.Recursion/2.Subsequences Pattern/python/01.Generate all valid parenthesis.py
"""
=============================================================================
  QUESTION: 22. Generate Parentheses (LeetCode)
=============================================================================

  Given n pairs of parentheses, generate all combinations of well-formed parentheses.
  Example: n=3 → ["((()))","(()())","(())()","()(())","()()()"]

  APPROACH: Recursion with Open/Close Count
  - Add '(' if open < n. Add ')' ONLY if close < open.
  - Base: open == n and close == n → valid string found.

  Recursion tree for n=2:
                    ""
                    |
                   "("
                 /      \\
             "(("        "()"
            /    \\          |
        "(()"   INVALID   "()(""
          |                  |
       "(())"             "()()"

  DRY RUN n=2:
    solve("",0,0) → solve("(",1,0) → solve("((",2,0) → solve("(()",2,1)
      → solve("(())",2,2) → PUSH "(())" ✓
    solve("(",1,0) → solve("()",1,1) → solve("()(",2,1) → solve("()()",2,2) → PUSH "()()" ✓

  Time: O(2^N × N), Space: O(N)
=============================================================================
"""


def solve(ans, s, open_count, close_count, n):
    if open_count == n and close_count == n:
        ans.append(s)
        return

    if open_count <= n:
        solve(ans, s + '(', open_count + 1, close_count, n)
    if open_count > close_count:
        solve(ans, s + ')', open_count, close_count + 1, n)


def generate_parenthesis(n):
    ans = []
    solve(ans, "", 0, 0, n)
    return ans


# Driver Code
if __name__ == "__main__":
    print(generate_parenthesis(3))
    # ["((()))","(()())","(())()","()(())","()()()"]

